File: bench/suite/src/st/fannkuch-redux/go/fannkuch-redux.go
// fannkuch-redux: single-threaded, same algorithm as the Benchmarks Game description
// (permutations generated in the reference order, checksum alternates sign).
package main

import (
	"fmt"
	"os"
	"strconv"
)

func fannkuch(n int) (int, int) {
	perm1 := make([]int, n)
	for i := range perm1 {
		perm1[i] = i
	}
	perm := make([]int, n)
	count := make([]int, n)
	maxFlips, checksum, permCount := 0, 0, 0
	r := n
	for {
		for ; r != 1; r-- {
			count[r-1] = r
		}
		copy(perm, perm1)
		flips := 0
		for k := perm[0]; k != 0; k = perm[0] {
			for i, j := 0, k; i < j; i, j = i+1, j-1 {
				perm[i], perm[j] = perm[j], perm[i]
			}
			flips++
		}
		if flips > maxFlips {
			maxFlips = flips
		}
		if permCount%2 == 0 {
			checksum += flips
		} else {
			checksum -= flips
		}
		// Next permutation.
		for {
			if r == n {
				return checksum, maxFlips
			}
			perm0 := perm1[0]
			for i := 0; i < r; i++ {
				perm1[i] = perm1[i+1]
			}
			perm1[r] = perm0
			count[r]--
			if count[r] > 0 {
				break
			}
			r++
		}
		permCount++
	}
}

func main() {
	n := 7
	if len(os.Args) > 1 {
		n, _ = strconv.Atoi(os.Args[1])
	}
	checksum, maxFlips := fannkuch(n)
	fmt.Printf("%d\nPfannkuchen(%d) = %d\n", checksum, n, maxFlips)
}

File: bench/suite/src/st/fannkuch-redux/java/fannkuchredux.java
// fannkuch-redux, single-threaded reference algorithm (Benchmarks Game description).
public final class fannkuchredux {
    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        int[] perm = new int[n], perm1 = new int[n], count = new int[n];
        int maxFlips = 0, checksum = 0, permCount = 0, r = n;
        for (int i = 0; i < n; i++) perm1[i] = i;
        outer:
        while (true) {
            while (r != 1) { count[r - 1] = r; r--; }
            System.arraycopy(perm1, 0, perm, 0, n);
            int flips = 0, k;
            while ((k = perm[0]) != 0) {
                for (int i = 0, j = k; i < j; i++, j--) { int t = perm[i]; perm[i] = perm[j]; perm[j] = t; }
                flips++;
            }
            if (flips > maxFlips) maxFlips = flips;
            checksum += (permCount & 1) == 0 ? flips : -flips;
            while (true) {
                if (r == n) break outer;
                int p0 = perm1[0];
                for (int i = 0; i < r; i++) perm1[i] = perm1[i + 1];
                perm1[r] = p0;
                if (--count[r] > 0) break;
                r++;
            }
            permCount++;
        }
        System.out.println(checksum + "\nPfannkuchen(" + n + ") = " + maxFlips);
    }
}
